### ISSUE TITLE
RestTemplate for eureka client failed to send heartbeat after eureka restart

### ISSUE THREAD

**Issue by Alvise88** [spring-cloud/spring-cloud-netflix#2861]
If in EDGWARE i remove jersey the possibility introduced to use RestTemplate has a different behaviour.   A client fails to send heart beat to eureka server after eureka restart. 
I think it depends on InstanceInfo or something linked with charsets

**Comment by ryanjbaxter** [comment]
Can you provide a project that reproduces the issue?

**Comment by ryanjbaxter** [comment]
Didnt sound like it to me, would need ore information about what the error is.

**Comment by maximusfloydus** [comment]
I also see this issue with Spring Boot 2.0.1.RELEASE and Spring Cloud Netflix 2.0.0.RC1.  It only happens when I am not using the jersey1 eureka client.

```
[2018-05-07 19:23:37:672] [WARN] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient] Request execution failed with message: 404 null
[2018-05-07 19:23:37:704] [ERROR] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient] Request execution error
org.springframework.web.client.HttpClientErrorException: 404 null
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:94)
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:79)
	at org.springframework.web.client.ResponseErrorHandler.handleError(ResponseErrorHandler.java:63)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:775)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:728)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:684)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:600)
	at org.springframework.cloud.netflix.eureka.http.RestTemplateEurekaHttpClient.sendHeartBeat(RestTemplateEurekaHttpClient.java:98)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient.executeOnNewServer(RedirectingEurekaHttpClient.java:118)
	at com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient.execute(RedirectingEurekaHttpClient.java:79)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:120)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.DiscoveryClient.renew(DiscoveryClient.java:846)
	at com.netflix.discovery.DiscoveryClient$HeartbeatThread.run(DiscoveryClient.java:1399)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[2018-05-07 19:23:37:704] [WARN] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient] Request execution failed with message: 404 null
[2018-05-07 19:23:37:704] [ERROR] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.DiscoveryClient] DiscoveryClient_NOT-YOURSERVICE/NOT-YOURSERVICE-af442b6e1d14 - was unable to send heartbeat!
com.netflix.discovery.shared.transport.TransportException: Retry limit reached; giving up on completing the request
	at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:139)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.DiscoveryClient.renew(DiscoveryClient.java:846)
	at com.netflix.discovery.DiscoveryClient$HeartbeatThread.run(DiscoveryClient.java:1399)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by ryanjbaxter** [comment]
@maximusfloydus IDK if this is the same issue, but can you provide a sample that reproduces the problem?

**Comment by maximusfloydus** [comment]
I will, in a few days...I hope.

**Comment by maximusfloydus** [comment]
@ryanjbaxter I tracked down the issue

[RestTemplateEurekaHttpClient.sendHeartBeat](https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClient.java#L98) needs to handle HttpClientErrorException thrown from DefaultResponseErrorHandler to get the status code and return a response with the status code.
```
[2018-06-19 11:26:10:023] [ERROR] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient]  
 + Message: Request execution error\n+ Throwable: org.springframework.web.client.HttpClientErrorException: 404 null
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:94)
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:79)
	at org.springframework.web.client.ResponseErrorHandler.handleError(ResponseErrorHandler.java:63)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:766)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:724)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:680)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:600)
	at org.springframework.cloud.netflix.eureka.http.RestTemplateEurekaHttpClient.sendHeartBeat(RestTemplateEurekaHttpClient.java:98)
```


I tested locally using Eureka 1.9.2 - using RestTemplate for Eureka Client vs Jersey

State 1:

- Eureka server UP
- Client application UP
- Client application registiers to Eureka Server
- client application heartbeats fine

State 2:

- kill eureka server
- client application errors trying to heartbeats
- wait a few minutes

State 3:

- Start eureka server UP
- client application errors with 404 error on heartbeat

Eureka Service logs
```
PUT /eureka/apps/MYAPP/MYAPP-myhostname?status=UP&lastDirtyTimestamp=1529425763653 HTTP/1.1" 404
[WARN] [http-nio-8080-exec-4] [com.netflix.eureka.resources.InstanceResource] Not Found (Renew): MYAPP - MYAPP-myhostname
```

**Comment by maximusfloydus** [comment]
Can you also add it to the next 2.0.x release?

**Comment by ryanjbaxter** [comment]
We need to look at it.  It will be fixed in the 1.4.x branch and then merged forward into the 2.0.x branch to be picked up in the next release.

**Comment by ryanjbaxter** [comment]
Fixed via 7b75fe9
